File: BST/BinarySearchTree.java
public class BinarySearchTree {
	private class TreeNode {
		public char val;
		public TreeNode left;
		public TreeNode right;

		public TreeNode(char ch) {
			this.val = ch;
		}
	}

	private TreeNode root;

	public boolean search(char target) {
		return search(target, root);
	}

	private boolean search(char target, TreeNode node) {
		if (node != null) {
			if (node.val == target) {
				return true;
			} else if (node.val > target) {
				return search(target, node.left);
			} else {
				return search(target, node.right);
			}
		}
		return false;
	}

	public void insert(char ch) {
		root = insert(ch, root);
	}

	private TreeNode insert(char ch, TreeNode node) {
		if (node == null) {
			return new TreeNode(ch);
		} else if (node.val > ch) {
			node.left = insert(ch, node.left);
		} else if (node.val < ch) {
			node.right = insert(ch, node.right);
		}
		return node;
	}
}

File: BST/BinarySearchTree.c
#include <stdlib.h>
#include <stdio.h>

#define FALSE 0;
#define TRUE 1;

typedef struct TreeNode_st {
	char val;
	struct TreeNode_st* left;
	struct TreeNode_st* right;
} TreeNode, *TreeNodePtr;

typedef struct BinarySearchTree_st {
	TreeNodePtr root;
} BinarySearchTree, *BinarySearchTreePtr;

int Search_BST(TreeNodePtr node_ptr, char ch);

TreeNodePtr Insert_BST(TreeNodePtr node_ptr, char ch);

TreeNodePtr Construct_TreeNode(char ch) {
	TreeNodePtr res = (TreeNodePtr)malloc(sizeof(*res));
	return res;
}

BinarySearchTreePtr Construct_BST() {
	BinarySearchTreePtr res = (BinarySearchTreePtr)malloc(sizeof(*res));
	return res;
}

int Search(BinarySearchTreePtr bst_ptr, char ch) {
	int result = Search_BST(bst_ptr->root, ch);
	return result;
}

int Search_BST(TreeNodePtr node_ptr, char ch) {
	if (node_ptr == NULL) {
		return FALSE;
	} else {
		if (node_ptr->val > ch) {
			Search_BST(node_ptr->left, ch);
		} else if (node_ptr->val < ch) {
			Search_BST(node_ptr->right, ch);
		} else {
			return TRUE;
		}
		return FALSE;
	}
}

void Insert(BinarySearchTreePtr bst_ptr, char ch) {
	bst_ptr->root = Insert_BST(bst_ptr->root, ch);
}

TreeNodePtr Insert_BST(TreeNodePtr node_ptr, char ch) {
	if (node_ptr == NULL) {
		return Construct_TreeNode(ch);
	} else if (node_ptr->val > ch) {
		node_ptr->left = Insert_BST(node_ptr->left, ch);
	} else if (node_ptr->val < ch) {
		node_ptr->right = Insert_BST(node_ptr->right, ch);
	}
	return node_ptr;
}

void FreeBST(TreeNodePtr node_ptr) {
	if (node_ptr != NULL) {
		FreeBST(node_ptr->left);
		FreeBST(node_ptr->right);
		free(node_ptr);
	}
}

void FreeBinarySearchTree(BinarySearchTreePtr bst_ptr) {
	FreeBST(bst_ptr->root);
}

int main(int args, char** argv) {
	BinarySearchTreePtr bst_ptr = Construct_BST();
	printf("Successfully construct\n");
	Insert(bst_ptr, 'm');
	printf("Successfully insert\n");
	Insert(bst_ptr, 'h');
	Insert(bst_ptr, 'r');
	Insert(bst_ptr, 'c');
	Insert(bst_ptr, 'p');
	FreeBinarySearchTree(bst_ptr);
}